### PR TITLE
Adding config to specify the type of replication model to be used for inter colo replication

### PR DESCRIPTION
**Pull request by Arun-LinkedIn**
Adding config to specify the type of replication to be used for inter colo replication. It has two values - "all-to-all"(current model) or "leader-based" with default value is "all-to-all". If set to "leader-based", inter colo replication will be limited to the leaders(as elected by Helix) of the partition of each datacenter. If set to "all-to-all", inter colo replication will be in an all-to-all fashion, i.e. each replica talks to all other replicas irrespective of their state. Intra colo replication will continue as all-to-all fashion in both the models.

### CODE REVIEW

File: ambry-api/src/main/java/com/github/ambry/replication/ReplicationModelType.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package com.github.ambry.replication;
+
+/**
+ * The type of replication used for inter-data centers
```
Comment (jsjtzyy) [inline ambry-api/src/main/java/com/github/ambry/replication/ReplicationModelType.java:18]: nit: `The model used for cross-colo replication`

File: ambry-replication/src/main/java/com/github/ambry/replication/ReplicationManager.java [hunk 2]
```diff
@@ -254,6 +254,14 @@ private void updatePartitionInfoMaps(List<RemoteReplicaInfo> remoteReplicaInfos,
     return Collections.unmodifiableMap(peerLeaderReplicasByPartition);
   }
 
+  /**
+   * Get a replication model type being used for inter colo replication
+   * @return replication model type {@link ReplicationEngine#replicationModelType}
+   */
+  ReplicationModelType  getReplicationModelType(){
```
Comment (jsjtzyy) [inline ambry-replication/src/main/java/com/github/ambry/replication/ReplicationManager.java:261]: nit: format this file

File: ambry-replication/src/test/java/com/github/ambry/replication/ReplicationTest.java [hunk 3]
```diff
@@ -171,6 +171,38 @@ public ReplicationTest(short requestVersion, short responseVersion) throws Excep
     replicationConfig = new ReplicationConfig(verifiableProperties);
   }
 
+  /**
+   * Tests replication model is correct obtained from properties
+   * @throws Exception
+   */
+  @Test
+  public void verifyReplicationTypeFromConfig() throws Exception {
```
Comment (jsjtzyy) [inline ambry-replication/src/test/java/com/github/ambry/replication/ReplicationTest.java:179]: minor: to make the naming consistent, how about renaming this test to `replicationTypeFromConfigTest`?

File: ambry-replication/src/test/java/com/github/ambry/replication/ReplicationTest.java [hunk 4]
```diff
@@ -171,6 +171,38 @@ public ReplicationTest(short requestVersion, short responseVersion) throws Excep
     replicationConfig = new ReplicationConfig(verifiableProperties);
   }
 
+  /**
+   * Tests replication model is correct obtained from properties
```
Comment (jsjtzyy) [inline ambry-replication/src/test/java/com/github/ambry/replication/ReplicationTest.java:175]: nit:  `is correctly obtained`

File: ambry-api/src/main/java/com/github/ambry/config/ReplicationConfig.java [hunk 5]
```diff
@@ -181,6 +181,17 @@
...
 
+  /**
+   * To specify the type of replication to be used for inter colo replication.
+   * It has two values - "all-to-all" or "leader-based". Default value is "all-to-all".
+   * If set to "leader-based", inter colo replication between will be limited to the leaders(as elected by Helix) of the partition of each datacenter.
+   * If set to "all-to-all", inter colo replication will be in an all-to-all fashion, i.e. each replica talks to all other replicas irrespective of their state.
+   * Intra colo replication will continue as all-to-all fashion in both the models.
+   */
+  @Config("replication.model.across.datacenters")
+  @Default("all-to-all")
```
Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com/github/ambry/config/ReplicationConfig.java:192]: nit: I suggest create a static string variable for this one
```
public static final String DEFAULT_REPLICATION_MODEL_ACROSS_DATACENTRS = ReplicationModelType.ALL_TO_ALL.toString().toLowerCase();

@Config("replication.model.across.datacenters")
@Default(DEFAULT_REPLICATION_MODEL_ACROSS_DATACENTERS)
public final String replicationModelAcrossDatacenters;
```

File: ambry-api/src/main/java/com/github/ambry/replication/ReplicationModelType.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+/**
+ * Copyright 2019 LinkedIn Corp. All rights reserved.
```
Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com/github/ambry/replication/ReplicationModelType.java:2]: the year should be 2020.

File: ambry-api/src/main/java/com/github/ambry/config/ReplicationConfig.java (hunk 5 above)
Comment (Arun-LinkedIn) [reply ambry-api/src/main/java/com/github/ambry/config/ReplicationConfig.java]: Sure, added a static string constant variable. I directly defined the variable name as - public static final String DEFAULT_REPLICATION_MODEL_ACROSS_DATACENTRS = "all-to-all". It wasn't accepting enum name as a constant variable in the @default annotation.

Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com/github/ambry/config/ReplicationConfig.java]: You don't have to pass enum name to the @Default annotation. My idea is to make sure we have consistent string representation here. If we assign DEFAULT_REPLICATION_MODEL_ACROSS_DATACENTERS with the enum's lowercase string, then in ReplicationEngine you can just do
```
replicationModelType = ReplicationModelType.valueof(replicationConfig.replicationModelAcrossDatacenters.toUpperCase());
```

File: ambry-api/src/main/java/com/github/ambry/config/ReplicationConfig.java [hunk 7]
```diff
@@ -181,6 +182,17 @@
   @Default("")
   public final Set<String> replicationVcrRecoveryPartitions;
 
+  /**
+   * To specify the type of replication to be used for inter colo replication.
+   * It has two values - "all-to-all" or "leader-based". Default value is "all-to-all".
```
Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com/github/ambry/config/ReplicationConfig.java:187]: nit: please update comment to reflect latest change. It should be "ALL_TO_ALL" and "LEADER_BASED".

**Pull request merged**